### ISSUE TITLE
AttributeError: 'Image' object has no attribute 'owner'

### ISSUE THREAD

**Issue by benyraja** [os-migrate/os-migrate#343]
When trying to export the workloads, getting this error. I have imported the images ( verified the owner id is present ), networks, keypairs, flavors etc..

`The full traceback is:
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1611234052.7762973-811256-75477694867635/AnsiballZ_export_workload.py", line 102, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1611234052.7762973-811256-75477694867635/AnsiballZ_export_workload.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1611234052.7762973-811256-75477694867635/AnsiballZ_export_workload.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.os_migrate.os_migrate.plugins.modules.export_workload', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/export_workload.py", line 135, in <module>
  File "/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/export_workload.py", line 131, in main
  File "/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/export_workload.py", line 121, in run_module
  File "/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py", line 60, in from_sdk
  File "/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/resource.py", line 90, in from_sdk
  File "/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py", line 171, in _refs_from_sdk
  File "/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/reference.py", line 29, in image_ref
  File "/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/reference.py", line 285, in _fetch_ref
  File "/usr/lib/python3.6/site-packages/openstack/resource.py", line 586, in __getattribute__
    raise e
  File "/usr/lib/python3.6/site-packages/openstack/resource.py", line 580, in __getattribute__
    return object.__getattribute__(self, name)
AttributeError: 'Image' object has no attribute 'owner'
failed: [localhost] (item={'name': 'benytestmigration2', 'id': '44eedeef-586f-40a0-9c6d-59f252aaad27'}) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": {
        "id": "44eedeef-586f-40a0-9c6d-59f252aaad27",
        "name": "benytestmigration2"
    },
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1611234052.7762973-811256-75477694867635/AnsiballZ_export_workload.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1611234052.7762973-811256-75477694867635/AnsiballZ_export_workload.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1611234052.7762973-811256-75477694867635/AnsiballZ_export_workload.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.os_migrate.os_migrate.plugins.modules.export_workload', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/export_workload.py\", line 135, in <module>\n  File \"/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/export_workload.py\", line 131, in main\n  File \"/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/export_workload.py\", line 121, in run_module\n  File \"/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py\", line 60, in from_sdk\n  File \"/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/resource.py\", line 90, in from_sdk\n  File \"/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py\", line 171, in _refs_from_sdk\n  File \"/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/reference.py\", line 29, in image_ref\n  File \"/tmp/ansible_os_migrate.os_migrate.export_workload_payload_bz6h3007/ansible_os_migrate.os_migrate.export_workload_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/reference.py\", line 285, in _fetch_ref\n  File \"/usr/lib/python3.6/site-packages/openstack/resource.py\", line 586, in __getattribute__\n    raise e\n  File \"/usr/lib/python3.6/site-packages/openstack/resource.py\", line 580, in __getattribute__\n    return object.__getattribute__(self, name)\nAttributeError: 'Image' object has no attribute 'owner'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
`

Any directions to solve this would be helpful. I'm stuck on this for the past two days. 

Thanks,
Beny

**Comment by jistr** [comment]
Hello Beny,

i suspect this could be caused by a mismatched version of openstacksdk, different than what we test with (perhaps too old). From the stack trace it looks like you're not using virtualenv. Are you using openstacksdk from packages that come with your distro? To help us map the compatibility issue, can you post the output of `pip list | grep openstacksdk` and what distro are you using?

If my guess above about openstacksdk version issue is correct, then the easiest solution would probably be to run from virtualenv. I'm not sure how much experience with virtualenv you have, so i'll try to post some steps. I didn't test the exact commands below but i think they should be pretty close:

```
python3 -m venv /home/myuser/os_migrate_venv
source /home/myuser/os_migrate_venv/bin/activate
pip install --upgrade openstacksdk
pip install --upgrade 'ansible>=2.9.1,<2.10'

# retry running OS Migrate playbook here
```

And when running Ansible later, make sure to first run `source /home/myuser/os_migrate_venv/bin/activate` into the active shell session.

In case you want to explore more about how we prepare the environment we test the OS Migrate code with, you can look at:
https://github.com/os-migrate/os-migrate/blob/3bcf6a91e455d212fd2f173c2d3d5484628cd144/toolbox/build/build.sh
https://github.com/os-migrate/os-migrate/blob/3bcf6a91e455d212fd2f173c2d3d5484628cd144/toolbox/build/venv-requirements.txt

Jirka

**Comment by benyraja** [comment]
Hello Jirka, 

I'm using Redhat Openstack Platform 13 as my source environment and OSP 16 is my destination environment. The migrator VM is rhel8. I have installed the openstacksdk package from the yum repo. 
python3-openstacksdk-0.36.4-0.20200715054251.76d3b29.el8ost.noarch

Ansible version is 2.9.16

I have never used the python virtualenv before. But I will take look at the references. 

Thanks,
Beny

**Comment by jistr** [comment]
Thanks for the info. I think the migrator VM being RHEL 8, and openstacksdk coming from OSP 13 or newer, that's a scenario that we'd like to have working "out of the box" if possible.

If you don't mind continuing with virtualenv for the time being to get unblocked, please do. But i also consider this incompatibility a bug, and i will explore if we can rewrite our code to accommodate for a wider spectrum of openstacksdk versions, to have it working without virtualenv in the circumstances you described.

Thanks,
Jirka

**Comment by benyraja** [comment]
Thanks for considering it. When I try to use the virtualenv, the keystone endpoint SSL verification is failing.
[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed

Is there a way I specify the ansible to ignore the SSL or any other options to fix it?

Thanks,
Beny

**Comment by jistr** [comment]
Yes, i think these variables should help: https://github.com/os-migrate/os-migrate/issues/303#issuecomment-726860563

**Comment by benyraja** [comment]
Thanks the export_workload.yml worked fine. I'm facing similar type of issue in import_workload as well.  


`The full traceback is:
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1611312159.7801816-862119-48005581818750/AnsiballZ_import_workload_create_instance.py", line 102, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1611312159.7801816-862119-48005581818750/AnsiballZ_import_workload_create_instance.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1611312159.7801816-862119-48005581818750/AnsiballZ_import_workload_create_instance.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.os_migrate.os_migrate.plugins.modules.import_workload_create_instance', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py", line 255, in <module>
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py", line 251, in main
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py", line 241, in run_module
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py", line 76, in create
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py", line 86, in sdk_params
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py", line 197, in _refs_from_ser
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/reference.py", line 17, in image_id
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/reference.py", line 305, in _fetch_id
TypeError: find_image() got an unexpected keyword argument 'project_id'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1611312159.7801816-862119-48005581818750/AnsiballZ_import_workload_create_instance.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1611312159.7801816-862119-48005581818750/AnsiballZ_import_workload_create_instance.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1611312159.7801816-862119-48005581818750/AnsiballZ_import_workload_create_instance.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.os_migrate.os_migrate.plugins.modules.import_workload_create_instance', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py\", line 255, in <module>\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py\", line 251, in main\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py\", line 241, in run_module\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py\", line 76, in create\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py\", line 86, in sdk_params\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py\", line 197, in _refs_from_ser\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/reference.py\", line 17, in image_id\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_xsk2q5op/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/reference.py\", line 305, in _fetch_id\nTypeError: find_image() got an unexpected keyword argument 'project_id'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
`

There is no "project_id" argument found in the export_workloads roles or in the workloads.yml data directory. I'm still using the virtualenv as per your suggestion.

Thanks,
Beny

**Comment by jistr** [comment]
Thanks, that looks like a valid issue, i'll investigate. We do test workload migration with an instance booted from an image in CI, so i'll try to see what's wrong how come the tests didn't catch this.

**Comment by jistr** [comment]
I think i see where the bug is. We used a single project filtering mechanism when looking up resources, but the Glance API / SDK is different than the rest, in that it doesn't support additional filters (like project id) on the find_image method. This bug should only show up when the image is owned by the same tenant as the VM being migrated. Our CI presently tests with an instance booted from a public image, so it's not triggering this code path.

I'll work on a fix, but if you want to continue trying the workload migration in the meantime, you should be able to work around this by editing the exported `workloads.yml`. I think for the server you're trying the migrate, the image_ref will presently look something like:

```
   image_ref:
      domain_name: '%auth%'
      name: some_image_name
      project_name: '%auth%'
```

You can edit it like so, which i think should work around the issue:
```
   image_ref:
      domain_name: null
      name: some_image_name
      project_name: null
```

In case you use `boot_disk_copy: true` (it'sn not the default for servers booted from image in the source cloud, but it can be set explicitly and is a valid option here), i think while reading the code i just found another bug there. We try to look up the image in the destination cloud even though it's not used for anything when destination server is boot-from-volume. Due to our CI clouds all having the same public images prepopulated, this issue also does not cause a failure there. So if you want to run with `boot_disk_copy: true`, you would use this workaround instead:

```
   image_ref: null
```

The boot_disk_copy parameter is explained here: https://os-migrate.github.io/os-migrate/user/migration-params-guide.html

**Comment by benyraja** [comment]
Thanks for the help and suggestions, I was able to move forward those image_ref issues. 

During the snapshot creation, the disks more than 200G takes longer time to create and it is hitting a timeout. May I know what is the default timeout? and is there anyway we can increase it through parameters? 

```
The full traceback is:
Traceback (most recent call last):
  File "/root/.ansible/tmp/ansible-tmp-1611573678.58079-15786-103656024104673/AnsiballZ_import_workload_export_volumes.py", line 102, in <module>
    _ansiballz_main()
  File "/root/.ansible/tmp/ansible-tmp-1611573678.58079-15786-103656024104673/AnsiballZ_import_workload_export_volumes.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/root/.ansible/tmp/ansible-tmp-1611573678.58079-15786-103656024104673/AnsiballZ_import_workload_export_volumes.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.os_migrate.os_migrate.plugins.modules.import_workload_export_volumes', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python3.6/runpy.py", line 205, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py", line 537, in <module>
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py", line 533, in main
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py", line 525, in run_module
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py", line 296, in prepare_exports
  File "/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py", line 378, in _detach_data_volumes_from_source
  File "/home/myuser/os_migrate_venv/lib64/python3.6/site-packages/openstack/cloud/_block_storage.py", line 231, in create_volume
    "Timeout waiting for the volume to be available."):
  File "/home/myuser/os_migrate_venv/lib64/python3.6/site-packages/openstack/utils.py", line 68, in iterate_timeout
    raise exceptions.ResourceTimeout(message)
openstack.exceptions.ResourceTimeout: Timeout waiting for the volume to be available.
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1611573678.58079-15786-103656024104673/AnsiballZ_import_workload_export_volumes.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1611573678.58079-15786-103656024104673/AnsiballZ_import_workload_export_volumes.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1611573678.58079-15786-103656024104673/AnsiballZ_import_workload_export_volumes.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.os_migrate.os_migrate.plugins.modules.import_workload_export_volumes', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py\", line 537, in <module>\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py\", line 533, in main\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py\", line 525, in run_module\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py\", line 296, in prepare_exports\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload_1_xixmeo/ansible_os_migrate.os_migrate.import_workload_export_volumes_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_export_volumes.py\", line 378, in _detach_data_volumes_from_source\n  File \"/home/myuser/os_migrate_venv/lib64/python3.6/site-packages/openstack/cloud/_block_storage.py\", line 231, in create_volume\n    \"Timeout waiting for the volume to be available.\"):\n  File \"/home/myuser/os_migrate_venv/lib64/python3.6/site-packages/openstack/utils.py\", line 68, in iterate_timeout\n    raise exceptions.ResourceTimeout(message)\nopenstack.exceptions.ResourceTimeout: Timeout waiting for the volume to be available.\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```


Thanks,
Beny

**Comment by jistr** [comment]
Hello Beny,

the timeout is 10 minutes (600 seconds) and presently not configurable via a parameter. We will create a parameter for this, and consider a higher default value. Fortunately the value is defined in a single place, here:

https://github.com/os-migrate/os-migrate/blob/da4bc444d73417bc0df41b5a3bb3b5604cb4e27d/os_migrate/plugins/module_utils/workload_common.py#L11

So to work around the issue, you should be able to edit the value directly in your installed Ansible collection. The path of the file should be: `$HOME/.ansible/collections/ansible_collections/os_migrate/os_migrate/plugins/module_utils/workload_common.py`.

Jirka

**Comment by jistr** [comment]
Hello @benyraja,

we have released version 0.6.2 of the OS Migrate collection, with the fixes for bugs we discussed here, more info in changelog:

https://os-migrate.github.io/os-migrate/changelog.html#id1

It can be installed with:

```
ansible-galaxy collection install os_migrate.os_migrate:0.6.2
```

One more heads up: We discovered that openstacksdk 0.53 breaks `os_security_group_rule` module in Ansible (the one shipped with Ansible itself). The test suite didn't run as far to tell us whether that openstacksdk breaks OS Migrate as well, but i think it's a possibility. If you hit any strange issues, please make sure you're using openstacksdk 0.52 in your virtualenv. This can be ensured with:

```
pip install --upgrade 'openstacksdk>=0.52,<0.53'
```

We will still be looking into other issues like making workload migration timeouts configurable (so far we just increased the timeout from 10 to 30 minutes) and the compatibility with older openstack SDK versions.

**Comment by benyraja** [comment]
Hello Jirka,

Thanks for fixes. I'm facing another issue on the same os_migrate.os_migrate.import_workloads task. 

```
 File "/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_3uyjutm8/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py", line 136, in update_sdk_params_block_device_mapping
KeyError: 'image_id'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1611841212.1321933-235137-77393911168865/AnsiballZ_import_workload_create_instance.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1611841212.1321933-235137-77393911168865/AnsiballZ_import_workload_create_instance.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1611841212.1321933-235137-77393911168865/AnsiballZ_import_workload_create_instance.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.os_migrate.os_migrate.plugins.modules.import_workload_create_instance', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n    mod_name, mod_spec, pkg_name, script_name)\n  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_3uyjutm8/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py\", line 255, in <module>\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_3uyjutm8/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py\", line 251, in main\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_3uyjutm8/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/modules/import_workload_create_instance.py\", line 241, in run_module\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_3uyjutm8/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py\", line 82, in create\n  File \"/tmp/ansible_os_migrate.os_migrate.import_workload_create_instance_payload_3uyjutm8/ansible_os_migrate.os_migrate.import_workload_create_instance_payload.zip/ansible_collections/os_migrate/os_migrate/plugins/module_utils/server.py\", line 136, in update_sdk_params_block_device_mapping\nKeyError: 'image_id'\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1

```

In the workloads.yml file, i have updated the below params.

```
boot_disk_copy: true

image_ref: null
```

also, could you please explain what is the process behind the volume copy? I saw some references related to ndb. 

Thanks,
Beny

**Comment by jistr** [comment]
Hello Beny,

i've seen that error when working on one of the fixes, i'm thinking the workaround i advised earlier causes that place in code to error. I guess that you're still on version 0.6.1 where the code looks like this:

https://github.com/os-migrate/os-migrate/blob/0.6.1/os_migrate/plugins/module_utils/server.py#L135-L136

While in 0.6.2 i've added a check in that place and it shouldn't cause an error:

https://github.com/os-migrate/os-migrate/blob/0.6.2/os_migrate/plugins/module_utils/server.py#L135-L136

Can you try to updating your collection to 0.6.2 and see if it fixes your issue? If you want just these minimal fixes, please make sure to specify the exact version in the install command as i described in my previous comment. Shortly we may be releasing 0.7.0 which will also contain new features.

-----

Regarding the volume copy process, it's done via nbd over ssh.

In the source cloud:

* If boot_disk_copy is true, first a volume representing the boot disk is created. This differs in servers which were booted from image vs. booted from volume. If the server was booted from image, first a snapshot of the image is taken and it is converted to a volume. If it was booted from volume, a volume snapshot is created and it is converted to a separate volume.

* The boot volume is attached to the conversion host. Any other volumes which were attached to the server being migrated are temporarily detached from the server and attached to the conversion host too. They are all exported as NBD but on localhost only.

In the destination cloud:

* Blank volumes of matching sizes are created for all the volumes that will be migrated from the source cloud, and they are attached to the destination conversion host.

* The destination conversion host creates ssh tunnels (port forwards) for the source NBD exports, so that they can be accessed on localhost of the destination conversion host. (The destination conversion host had been authorized to ssh into the source conversion host during its deployment.)

* The data is copied from the NBD exports onto the blank volumes.

* The volumes are detached from the conversion host and ssh tunnels to the source are closed.

* A new server is created according to the specification from workloads.yml, and the volumes are attached to it. This is the migrated server. (The server is either boot-from-image or boot-from-volume, again depending on the boot_disk_copy parameter.)

In the source cloud:

* NBD exports are closed and volumes are detached from the conversion host and re-attached to the original server. (Skimming thorugh code now, i'm not sure if we delete the boot volume clone in the source cloud, i think we'll have to double check that and fix if necessary.)

-----

Have a good day,

Jirka

**Comment by jistr** [comment]
Hello @benyraja,

did you manage to get past the error we discussed last, and migrate some content/workloads between the clouds? I'm curious if/how the migration went further, and i would welcome any feedback.

Have a good day,

Jirka

**Comment by benyraja** [comment]
Hello Jirka,

Yes, the trail run was successful. some of my observations are

- The root disk sparsify didn't work as expected, ended up copying the 100G disk which has less than 10Gigs of data
- if something goes wrong after the volume disk copy (esp, during VM creation in the destination), the re-run starts the disk copy again from the start. It would be helpful, if it detects the disk that is already attached to the conversion host.
- I have migrated a VM with the name 'benymigration' and if the next VM name is 'beny', it exits the playbook as "VM already exists" message. Looks like it is validating VMs through name patterns and that may cause problems which uses name patterns. 

Some questions:
Is it possible to use RHEL8 as a conversion host? and what extra configs that we should do to make it work?
Is there a way to specify the VMs that needs migration/exclude from migration through the ansible parameter? rather than editing the workloads.yml file manually. 


I loved this tool, it will help us lot in the longer run. Your responses helped me a lot to setup this. Thank you so much. 

Beny

**Comment by jistr** [comment]
Thanks Beny, that's excellent feedback.

> Yes, the trail run was successful. some of my observations are
> 
>     * The root disk sparsify didn't work as expected, ended up copying the 100G disk which has less than 10Gigs of data
> 
>     * if something goes wrong after the volume disk copy (esp, during VM creation in the destination), the re-run starts the disk copy again from the start. It would be helpful, if it detects the disk that is already attached to the conversion host.
> 
>     * I have migrated a VM with the name 'benymigration' and if the next VM name is 'beny', it exits the playbook as "VM already exists" message. Looks like it is validating VMs through name patterns and that may cause problems which uses name patterns.

I've created GH issues for all three. Re avoiding repeated copy on failed migration, there are edge cases we'd have to watch out for in order to make such solution safe (i wrote further details at https://github.com/os-migrate/os-migrate/issues/366#issuecomment-776053930 ). So i'm not sure about that one, whether we want to try the reuse or rather clean up better after a failed migration. The other two sound like bugs we should investigate with priority.

> 
> Some questions:
> Is it possible to use RHEL8 as a conversion host? and what extra configs that we should do to make it work?

Yes, but that's quite recent work and on 0.6.x i wouldn't call it usable. There have been several improvements on the main branch (unregistration, and more variables to control registration) which aren't yet released into Ansible Galaxy. So i would recommend trying RHEL only with the upcoming OS Migrate 0.7. I was hoping we could look into e.g. the compatibility issues with older OpenStack SDK before releasing 0.7, as fixing those might potentially cause some small, but backwards-incompatible, changes in resource serialization.

Re extra config when using RHEL, i created a GH issue for documenting that. The image we tested with is RHEL 8.3 KVM Guest Image qcow2 downloaded from the Customer Portal. The hosts need to register themselves to install a few extra packages, registration is configured via these variables (once 0.7.0 comes out): https://github.com/os-migrate/os-migrate/blob/93bcd27b044ad15c818bdd3fd86a9c089776a608/os_migrate/roles/conversion_host_content/tasks/rhel.yml#L5-L27

> Is there a way to specify the VMs that needs migration/exclude from migration through the ansible parameter? rather than editing the workloads.yml file manually.

We have filter variables which work **during export**:
https://os-migrate.github.io/os-migrate/user/variables-guide.html#resource-filters

So these allow to export a subset of the resources into YAML, and then the import playbook acts on the whole YAML. If you meant the inverse approach of exporting all workloads metadata into YAML, and then filtering them when running the import playbook instead -- i think that would make sense as well. Filtering during import isn't presently implemented but shouldn't be difficult to implement, i created a GH issue.

> 
> I loved this tool, it will help us lot in the longer run. Your responses helped me a lot to setup this. Thank you so much.

Great! Thank you for helping us improve the tool with your feedback.

Jirka

**Comment by jistr** [comment]
Hello @benyraja,

we just released version 0.7 with the RHEL conversion host support, RHEL variables are listed here: https://os-migrate.github.io/os-migrate/user/variables-guide.html#conversion-host-rhel-variables

In my previous post i forgot to mention that in addition to the registration variables, RHEL also needs this:

```
os_migrate_conversion_host_ssh_user: cloud-user
```

The bugs in sparsification and VM name matching should now be fixed. In my dev env i only had a 5 GB volume to test the sparsification on, but i did see a performance improvement there, so i'm hoping it works correctly and it would result in a considerable speedup in your 100 GB total / 10 GB used scenario. Any existing conversion hosts need to be redeployed for the sparsification fix to take effect.

OS Migrate now also attempts to auto-create floating IPs on interfaces where the source VM had floating IPs. This behavior can be switched off via `floating_ip_mode` migration parameter in the resource YAML: https://os-migrate.github.io/os-migrate/user/migration-params-guide.html#workload-migration-parameters
